### PR TITLE
Make isDigit branchless

### PR DESCRIPTION
**Pull request by UplinkCoder**
It results in smaller code which is always good

### CODE REVIEW

File: std/ascii.d [hunk 1]
```diff
@@ -314,7 +314,7 @@ bool isUpper(dchar c) @safe pure nothrow @nogc
   +/
 bool isDigit(dchar c) @safe pure nothrow @nogc
 {
-    return '0' <= c && c <= '9';
+    return (cast(ubyte)(c - '0')) < 10;
```
Comment (thewilsonator) [inline std/ascii.d:317]: this would give incorrect results for `cast(dchar)('0'+256)`

Comment (UplinkCoder) [reply std/ascii.d:317]: uh did I forget to change it to uint?
let me fix that

**Comment by UplinkCoder** [comment]
@thewilsonator can you put an auto-merge-squash?

File: std/ascii.d [hunk 2]
```diff
@@ -30,7 +30,7 @@ $(TR $(TD Validation) $(TD
         $(LREF isWhite)
 ))
 $(TR $(TD Conversions) $(TD
-        $(LREF toLower)
+        $(LREF toLower)
```
Comment (thewilsonator) [inline std/ascii.d:33]: no

**Review by thewilsonator — APPROVED** [review]

File: std/ascii.d (hunk 2 above)
Comment (UplinkCoder) [reply std/ascii.d:33]: damn that webinterface. fixing

**Comment by maxhaton** [comment]
On which compiler? https://d.godbolt.org/z/fMfjjYcaM code is actually slightly longer on ldc. 

https://d.godbolt.org/z/TeGq5Y1x6 The way that they don't get inlined is probably a more pressing issue

**Comment by UplinkCoder** [comment]
@maxhaton i fixed the code. try it again.

**Comment by maxhaton** [comment]
> @maxhaton i fixed the code. try it again.

https://d.godbolt.org/z/noh54odGo for reference. Was smaller code in reference to the source or binary?

**Comment by UplinkCoder** [comment]
@maxhaton it's smaller for unoptimized builds and dmd.

**Comment by thewilsonator** [comment]
I haven't look at the generated code, but in general we shouldn't be pandering for DMD at the expense of LDC an GDC. I understand this is useful for unoptimised build.

**Comment by UplinkCoder** [comment]
I could also put a micro-optimisation pass into dmd to transform (v <= C1 && v >= C2) into ((v - C1) > (C1 - C2))

**Comment by UplinkCoder** [comment]
There is no expense.
ldc and gdc to the rewrite I made manually automatically.
they generate the same code with and without this patch.

**Comment by maxhaton** [comment]
> I could also put a micro-optimisation pass into dmd to transform (v <= C1 && v >= C2) into ((v - C1) > (C1 - C2))

I guess you could, but it's probably just more spaghetti in the backend in exchange for a speedup that most people interested wouldn't even notice because optimized dmd builds are borderline considered harmful next to LLVM and GNU

**Review by RazvanN7 — COMMENTED** [review]
This probably needs a rebase to fix the buildkite failure.

**Comment by UplinkCoder** [comment]
I can't rebase it I think because I don't know where the branch lives

**Comment by thewilsonator** [comment]
it lives https://github.com/dlang/phobos/tree/UplinkCoder-patch-1

**Comment by berni44** [comment]
See #7800 for a similar PR that was refuted.

**Comment by dukc** [comment]
This is going to be incompatible with using some third-party overflow checking tools. Is that a problem?

**Comment by UplinkCoder** [comment]
Seeing the critique in the previous PR.
I have to agree that this is unreadable code.

**Comment by ljmf00** [comment]
> Seeing the critique in the previous PR.
> I have to agree that this is unreadable code.

The readability can be fixed with a comment. Also, musl libc implements it that way.

**Comment by ljmf00** [comment]
It is also interesting to see how a PR move forward depending on the people :eyes:

**Pull request closed without merging**